### ISSUE TITLE
How does RPC work with directory separators?

### ISSUE THREAD

**Issue by equeim** [transmission/transmission#3915]
When using RPC methods that accept file/directory paths, what directory separators should client use when Transmission is running on Windows? I found out that both native (`\`) and Unix (`/`) separators work, although what Transmission returns may be inconsistent.

E.g. if you call `torrent-set-location` with Unix separators, Transmission will accept that and then return `torrent-get` with `downloadDir` for this torrent that uses Unix separators too (even if it previously returned it using native separators), That's a bit odd.

I think it would be a good idea to document what RPC clients should do in this case.

**Comment by ckerr** [comment]
Good call. The answer is `/` but off the top of my head I don't think that's documented anywhere. @equeim do you want to make a PR for `docs/rpc-spec.md`?

**Comment by ckerr** [comment]
Hmm this might not be just a documentation question? I need to look at the RPC code and see if Windows separators will get converted into Unix ones or not. Assigning this to me since that may be out-of-scope for a documentation PR

**Comment by equeim** [comment]
Just checked this with Transmission 3.0.0 on Windows 11.

If I add torrent using official transmission-qt GUI then torrent-get/downloadDir will have Unix directory separators.

However, default value of session-get/download-dir uses Windows directory separators, and if RPC client uses it as torrent-add/download-dir then torrent-get/downloadDir will have Windows directory separators too.

**Comment by equeim** [comment]
BTW it would be nice to know whether transmission-daemon is running on Windows, so that client could convert all paths to uniform representation internally (I didn't see any such property in rpc-spec).

**Comment by killemov** [comment]
@equeim Most software use "/" internally no matter the platform. It's usually the (Windows) OS-calls that need translations for paths. You also want to abstract the software from the OS as much as possible. See also: https://superuser.com/questions/176388/why-does-windows-use-backslashes-for-paths-and-unix-forward-slashes

**Comment by equeim** [comment]
I agree, but among my concerns is also UI. If transmission-daemon is running on Windows and RPC client on Linux, I want to display correct paths (with \ separator) in its UI. For that I need transmission-daemon to tell me on which OS it is running.

**Comment by killemov** [comment]
> @killemov I agree, but among my concerns is also UI. If transmission-daemon is running on Windows and RPC client on Linux, I want to display correct paths (with \ separator) in its UI. For that I need transmission-daemon to tell me on which OS it is running.

Apparently you don't agree. ( You also want to abstract the software from the OS as much as possible. ) But I understand if you want to use paths in the UI for server-side processing.

**Comment by equeim** [comment]
I'm confused how you came to this conclusion.
I want to use Unix paths for RPC communication with transmission-daemon (if it is indeed what transmission-daemon wants) and native paths in UI.
For context, I'm talking specifically about paths that are only relevant to the server side, such as torrent's download directory. For that I need an ability co convert between Unix and native paths and back. This requires that client knows what "native" path is. Since OS on which client is running is irrelevant in this case, it needs to know on which OS the server is running.

Another case is user input. If server is running on Windows then I need client to be able Windows paths from user and convert them to Unix paths internally. However I cannot blindly replace all `\` characters in path with '/' because this will break when server is running on Linux (Linux file names are allowed co contain `\` character). Again, I need information about server's OS.

**Comment by mikedld** [comment]
I'm slightly confused. Why do you want to transform paths entered by the user in any way? Users know what OS is running on a remote server and know what format to use for paths they specify. If the issue is that daemon running on Windows returns you paths with forward slashes via RPC then I think we may want to fix the daemon (or rather libtransmission), where it makes sense.

**Comment by ckerr** [comment]
Looking over the 4.0.0 milestone issues, I realize that I've been leaving this one untouched because I just don't care enough about this issue to do the work on it. The current setup's been working [since 2010](https://github.com/transmission/transmission/blob/main/news/news-2.42-and-older.md#transmission-190-2010-02-16) so I think I was wrong to mark this as blocking 4.0.0.

I'm leaving this ticket open for now because I'm happy to do a PR review if someone else wants to make a PR for this.
